driver/src/price_estimation/kraken/api.rs

```diff
@@ -0,0 +1,265 @@
+}
+
+/// An HTTP Kraken API Client.
+#[derive(Debug)]
+pub struct KrakenHttpApi {
+    /// The base URL for the API calls.
+    base_url: String,
+    /// An HTTP client for all of the HTTP requests.
+    client: HttpClient,
+}
+
+/// The default Kraken API base URL.
+const DEFAULT_API_BASE_URL: &str = "https://api.kraken.com/0/public";
+
+impl KrakenHttpApi {
+    pub fn new() -> Result<Self> {
+        KrakenHttpApi::with_url(DEFAULT_API_BASE_URL)
+    }
+
+    pub fn with_url(base_url: &str) -> Result<Self> {
+        let client = HttpClient::new().context("failed to initialize HTTP client")?;
+        Ok(KrakenHttpApi {
+            base_url: base_url.into(),
+            client,
+        })
+    }
+}
+
+impl KrakenApi for KrakenHttpApi {
+    fn assets(&self) -> Result<HashMap<String, Asset>> {
+        self.client
+            .get(format!("{}/Assets", self.base_url))
+            .context("failed to retrieve list of assets from Kraken")?
+            .json::<KrakenResult<_>>()
+            .context("failed to parse assets JSON")?
+            .into_result()
+    }
+
+    fn asset_pairs(&self) -> Result<HashMap<String, AssetPair>> {
+        self.client
+            .get(format!("{}/AssetPairs", self.base_url))
+            .context("failed to retrieve list of asset pairs from Kraken")?
+            .json::<KrakenResult<_>>()
+            .context("failed to parse asset pairs JSON")?
+            .into_result()
+    }
+
+    fn ticker(&self, pairs: &[&str]) -> Result<HashMap<String, TickerInfo>> {
+        self.client
+            .get(format!("{}/Ticker?pair={}", self.base_url, pairs.join(",")))
+            .context("failed to retrieve ticker infos from Kraken")?
+            .json::<KrakenResult<_>>()
+            .context("failed to parse ticker JSON")?
+            .into_result()
+    }
+}
+
+/// The result type that is returned by Kraken on API requests. This type is
+/// only used internally.
+#[derive(Clone, Debug, Deserialize)]
+struct KrakenResult<T> {
+    error: Vec<String>,
+    result: Option<T>,
+}
+
+impl<T> KrakenResult<T> {
+    fn into_result(self) -> Result<T> {
+        if let Some(result) = self.result {
+            Ok(result)
+        } else if !self.error.is_empty() {
+            Err(anyhow!("Kraken API errors: {:?}", self.error))
+        } else {
+            Err(anyhow!("unknown Kraken API error"))
+        }
+    }
+}
+
+/// A struct representing an asset retrieved from the Kraken API.
+///
+/// Note that this is only a small subset of the data provided by the Kraken API
+/// and only the parts required for retrieving price estimates for the solver
+/// are included.
+#[derive(Clone, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct Asset {
+    pub altname: String,
+}
+
+impl Asset {
+    /// Create a new asset from an alternate name.
+    #[cfg(test)]
+    pub(crate) fn new(altname: &str) -> Asset {
```

Comment:
What is the purpose of `pub(crate)` here? No need to comment that in the code. I would just like to know.

Reply:
Oh! that was probably a copy pasta booboo, I'll fix it.